### ISSUE TITLE
PHPUnit\Framework\TestCase::__constructor marked as internal even though it's intended to be used

### ISSUE THREAD

**Issue by dkarlovi** [sebastianbergmann/phpunit#3975]
| Q                   | A
| --------------------| ---------------
| PHPUnit version     | 8.5.0
| PHP version         | 7.3.12
| Installation Method | PHAR

#### Summary

Method `PHPUnit\Framework\TestCase::__constructor` is marked as internal with this message:

```
@internal This method is not covered by the backward compatibility promise for PHPUnit
```

#### Current behavior

Method is marked as internal, seemingly just to signal it not being covered by BC.

But it's not actually internal since it's intended to be used by the fast you extend `TestCase` in your own suites.

The usage of internal methods is reported by running Psalm on your tests:

```
ERROR: InternalMethod - tests/Xezilaires/Bridge/PhpSpreadsheet/RowIteratorTest.php:28:37 - The method PHPUnit\Framework\TestCase::__construct has been marked as internal
final class RowIteratorTest extends TestCase
```

#### How to reproduce

Run Psalm on your tests backed by PHPUnit 8.5.0.

#### Expected behavior

Method(s) which are intended to be used by the tests shouldn't be marked as internal.

**Comment by dkarlovi** [comment]
Just found another one:

```
ERROR: InternalMethod - tests/Xezilaires/SpreadsheetIteratorTest.php:123:30 - The method PHPUnit\Framework\MockObject\Builder\InvocationMocker::willReturn has been marked as internal
                    $mocker->willReturn($spec['return']);
```

**Comment by sebastianbergmann** [comment]
I do not know where you get the idea that `PHPUnit\Framework\TestCase::__constructor()` is intended to be used in your own tests.

**Comment by dkarlovi** [comment]
It's not used directly, but the ctor is invoked just by extending `TestCase`, no? So basically any test case out there is using an internal method by default.

**Comment by r-sweet-codes** [comment]
Just as dkarlovi said, how am I supposed to extend the Testcase, if the extended class needs its own constructor (and thus parent::constructor(); is unavoidable)

**Comment by JohnRDOrazio** [comment]
After upgrading PHPUnit from 9 to 10, I started getting errors about not having enough arguments for the constructor, when trying to extend `TestCase` and then instantiating my class. Only way I could get around the error was to explicitly declare a constructor, which I wasn't even using before:
```php
class MyTest extends TestCase {

    public function __construct($name = '') {
        parent::__construct($name);
    }

    public function test() {
        //do a test
    }

}
```

Why is the constructor required? Why can't the `$name` parameter just default to an empty or null value in `TestCase` like it used to?

**Comment by manuirrazabal** [comment]
I have exactly the same error, after I upgrade from 9 to 10, my test cases do not have construct, therefore Im not extending. I dont understand why @sebastianbergmann keep closing this issue. Is there a proper documentation to upgrade properly. or we do have to do the workaround that @JohnRDOrazio did ?

**Comment by TomasVotruba** [comment]
Related change of `TestCase::__construct()` https://github.com/sebastianbergmann/phpunit/commit/705874f1b867fd99865e43cb5eaea4e6d141582f, first shipped in PHPUnit 10.0.0

I'm adding this rule to Rector upgrade set, so PHPUnit upgrade is seamless 👍
